File: 200-number-of-islands/number-of-islands.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        rows = len(grid)
        cols = len(grid[0])
        total_islands = 0

        for i in range(rows):
            for j in range(cols):
                if grid[i][j] != "0":
                    self.dfs(i, j, rows, cols, grid)
                    total_islands += 1
        return total_islands


    def dfs(self, row, col, rows, cols, grid):
        if row < 0 or row >= rows or col < 0 or col >= cols or grid[row][col] != "1":
            return
        
        grid[row][col] = "0"
        self.dfs(row-1, col, rows, cols, grid)
        self.dfs(row+1, col, rows, cols, grid)
        self.dfs(row, col-1, rows, cols, grid)
        self.dfs(row, col+1, rows, cols, grid)

File: 200-number-of-islands/number-of-islands.java
class Solution {
    public int numIslands(char[][] grid) {
        int rows = grid.length;
        int cols = grid[0].length;
        int result = 0;
        for(int i = 0; i < rows; i++) {
            for(int j = 0; j < cols; j++) {
                if(grid[i][j] != '0') {
                    dfs(i, j, rows, cols, grid);
                    result += 1;
                }
            }
        }
        return result;
    }

    private void dfs(int row, int col, int rows, int cols, char[][] grid) {
        if(row < 0 || row >= rows || col < 0 || col >= cols || grid[row][col] != '1')
            return;

        grid[row][col] = '0';
        dfs(row - 1, col, rows, cols, grid);
        dfs(row + 1, col, rows, cols, grid);
        dfs(row, col - 1, rows, cols, grid);
        dfs(row, col + 1, rows, cols, grid);
    }
}